### DISCUSSION TITLE
Map query params to controller arguments with Attributes

### DISCUSSION THREAD

**Discussion by mortenscheel** [laravel/framework#52185] [Ideas]
I just noticed that Symfony has the ability to inject query params as controller arguments by using attributes: https://symfony.com/doc/current/controller.html#mapping-query-parameters-individually

Having no experience whatsoever with how dependency injection works in Laravel, I obviously had to give it a try 😆. 
Here's the proof of concept I came up with. I'm *not* claiming that this is the way to do it, I just wanted to hear if you guys have any thoughts. Wouldn't this be a nice feature to have in Laravel? I haven't looked into it, but I suppose it might be possible to reuse some of the logic from Symfony.

In [Illuminate\Routing\ResolvesRouteDependencies](https://github.com/laravel/framework/blob/75d7b47191beeaa3cb3b59eeb5a3142561a9cb29/src/Illuminate/Routing/ResolvesRouteDependencies.php#L73-L89)
```php
protected function transformDependency(ReflectionParameter $parameter, $parameters, $skippableValue)
{
    $className = Reflector::getParameterClassName($parameter);
    // Hack start
    if (!$className && !empty($parameter->getAttributes(MapQueryParameter::class))) {
        if (($value = request()->query($parameter->getName())) !== null) {
            return match($parameter->getType()?->getName()){
                'int' => (int) $value,
                'float' => (float) $value,
                'bool' => (bool) $value,
                default => $value
            };
        }
    }
    // Hack end

    // If the parameter has a type-hinted class, we will check to see if it is already in
    // the list of parameters. If it is we will just skip it as it is probably a model
    // binding and we do not want to mess with those; otherwise, we resolve it here.
    if ($className && ! $this->alreadyInParameters($className, $parameters)) {
        $isEnum = (new ReflectionClass($className))->isEnum();

        return $parameter->isDefaultValueAvailable()
            ? ($isEnum ? $parameter->getDefaultValue() : null)
            : $this->container->make($className);
    }

    return $skippableValue;
}
```
And then inject query params in controller actions like this:
```php
public function index(#[MapQueryParameter] int $type)
{
    dd($type);
}
```
The argument type is optional, and you can give it a default value as usual.

EDIT: I've made a quick prototype package that adds this feature. It uses Symfony's `QueryParameterValueResolver` to resolve, validate and cast the values. https://github.com/mortenscheel/inject-query-params

**Comment by ghost** [comment]
Hi. Can you please explain the gains of doing this?

Why 
        
        public function index(Request $request): Response
        {
            dd($request->input('name');
        }

is not enough?

Why add overhead to the framework?

**Comment by mortenscheel** [comment]
> Why

It's certainly not a *necessary* feature. It's just a convenient alternative. And it's just a simple proof of concept. If you look at the Symfony link you can see that there are lots of other possibilities. Like mapping the query parameters to a DTO.

Regarding overhead, I'm pretty sure it's negligible.

**Reply by ghost** [reply]
Thank you. If it can be implemented as a package to not be included in the core, it would get my vote.

**Reply by mortenscheel** [reply]
Challenge accepted.
https://github.com/mortenscheel/inject-query-params

**Reply by ghost** [reply]
Congrats. Only one thing I see off in the readme of your lib. Validation exception. Query parameters are by definition optional. If you want to make them mandatory, you embed them in the route => you will get them in the controller function without attributes.

**Reply by mortenscheel** [reply]
Yeah that's one of the (many) things I wasn't sure how should be handled. Symfony throws a 404 which might actually make more sense, since required params sort or become part of the route signature. 

I'm still not sure if this is a feature I'd use myself, but it has been a great learning experience. And I could see it being added to the Laravel core at some point.

**Comment by mindaugasvcs** [comment]
Jealous thieves

**Reply by macropay-solutions** [reply]
@mindaugasvcs you can "steal" our auto filter for cruFd and implement it using Doctrine in Symfony like we implemented it via Eloquent.

**Comment by shaedrich** [comment]
Can you achieve this via [custom contextual attributes container binding](https://laravel.com/docs/12.x/container#defining-custom-attributes)?

**Comment by ejunker** [comment]
Here is a quick attempt at a `#[QueryParameter]` attribute based on the `RouteParameter` attribute. It supports casting to the correct type but you have to pass in the type you want. If someone can figure out how to use Reflection to detect the type let me know.

```php
<?php

namespace App\Http\Attributes;

use Attribute;
use Illuminate\Contracts\Container\Container;
use Illuminate\Contracts\Container\ContextualAttribute;

#[Attribute(Attribute::TARGET_PARAMETER)]
class QueryParameter implements ContextualAttribute
{
    /**
     * Create a new class instance.
     */
    public function __construct(
        public string $parameter,
        public string $type = 'default',
    ) {}

    /**
     * Resolve the route parameter.
     *
     * @return mixed
     */
    public static function resolve(self $attribute, Container $container, ?object $context = null)
    {
        $request = $container->make('request');

        // TODO can we use reflection to get the type?
        $resolveMethod = match ($attribute->type) {
            'int' => 'integer',
            'float' => 'float',
            'bool' => 'boolean',
            'string' => 'string',
            'date' => 'date',
            'enum' => 'enum',
            default => 'input',
        };

        return $request->{$resolveMethod}($attribute->parameter);
    }
}

```

Example usage:
```php
public function __invoke(Request $request, #[QueryParameter('posted', 'bool')] bool $posted = false)
{}
```

**Reply by shaedrich** [reply]
Why do you need the type twice (once as a string attribute argument value and once as a reflectable method argument type)?

**Reply by ejunker** [reply]
@shaedrich  Because I could not figure out how to get the type so I had to pass it in as an argument. If you can figure out how to get the type without having to pass it in as an argument that would be great.

**Reply by shaedrich** [reply]
I was actually working on a PR for this a while ago. I can look if I can publish this at least as a draft.

**Reply by macropay-solutions** [reply]
@ejunker what is the performance impact of this vs **$request->get(** or **->getQueryString(** or **->input(**? How do you handle exceptions for this?
